cache/handle.go

```diff
@@ -0,0 +1,425 @@
+func (r *Handle) Read(p []byte) (n int, err error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	var buf []byte
+
+	currentOffset := r.offset
+
+	buf, err = r.getChunk(currentOffset)
+	if err != nil && len(buf) == 0 {
+		return 0, io.EOF
+	}
+	readSize := copy(p, buf)
+	newOffset := currentOffset + int64(readSize)
+
+	r.offset = newOffset
+	if r.offset >= r.cachedObject.Size() {
+		return readSize, io.EOF
+	}
+
+	return readSize, err
+}
+
+// Close will tell the workers to stop
+func (r *Handle) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	close(r.preloadQueue)
+	go r.CacheFs().CleanUpCache(false)
+
+	fs.Debugf(r, "cache reader closed %v", r.offset)
+	return nil
+}
+
+// Seek will move the current offset based on whence and instruct the workers to move there too
+func (r *Handle) Seek(offset int64, whence int) (int64, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var err error
+	switch whence {
+	case os.SEEK_SET:
+		fs.Debugf(r, "moving offset set from %v to %v", r.offset, offset)
+		r.offset = offset
+	case os.SEEK_CUR:
+		fs.Debugf(r, "moving offset cur from %v to %v", r.offset, r.offset+offset)
+		r.offset += offset
+	case os.SEEK_END:
+		fs.Debugf(r, "moving offset end (%v) from %v to %v", r.cachedObject.Size(), r.offset, r.cachedObject.Size()+offset)
+		r.offset = r.cachedObject.Size() + offset
+	default:
+		err = errors.Errorf("cache: unimplemented seek whence %v", whence)
+	}
+
+	chunkStart := r.offset - (r.offset % r.CacheFs().chunkSize)
+	if chunkStart >= r.CacheFs().chunkSize {
+		chunkStart = chunkStart - r.CacheFs().chunkSize
+	}
+	r.queueOffset(chunkStart)
+
+	return r.offset, err
+}
+
+type worker struct {
+	r       *Handle
+	ch      <-chan int64
+	rc      io.ReadCloser
+	id      int
+	running bool
+	mu      sync.Mutex
+}
+
+// String is a representation of this worker
+func (w *worker) String() string {
+	return fmt.Sprintf("worker-%v <%v>", w.id, w.r.cachedObject.Name)
+}
+
+// reader will return a reader depending on the capabilities of the source reader:
+//   - if it supports seeking it will seek to the desired offset and return the same reader
+//   - if it doesn't support seeking it will close a possible existing one and open at the desired offset
+//   - if there's no reader associated with this worker, it will create one
+func (w *worker) reader(offset, end int64) (io.ReadCloser, error) {
+	var err error
+	r := w.rc
+	if w.rc == nil {
+		r, err = w.r.CacheFs().OpenRateLimited(func() (io.ReadCloser, error) {
+			return w.r.cachedObject.Object.Open(&fs.SeekOption{Offset: offset}, &fs.RangeOption{Start: offset, End: end})
+		})
+		if err != nil {
+			return nil, err
+		}
+		return r, nil
+	}
+
+	seekerObj, ok := r.(io.Seeker)
+	if ok {
+		_, err = seekerObj.Seek(offset, os.SEEK_SET)
+		return r, err
+	}
+
+	_ = w.rc.Close()
+	return w.r.CacheFs().OpenRateLimited(func() (io.ReadCloser, error) {
+		r, err = w.r.cachedObject.Object.Open(&fs.SeekOption{Offset: offset}, &fs.RangeOption{Start: offset, End: end})
```

Comment:
Good catch. I suspect the range option ended up being the final header otherwise I would have gotten a lot of bans with a really difficult bug to spot.

Hmm, I wasn't sure about the range support. It means I need a way to detect that and I don't know, disable cache completely for it? I don't think it makes sense anymore. I need to search for them and see what the impact is.